processing/src/main/java/io/druid/query/timeseries/TimeseriesQueryQueryToolChest.java

```diff
@@ -131,72 +136,95 @@ public TimeseriesQueryQueryToolChest(
       }
     };
 
-    return new QueryRunner<Result<TimeseriesResultValue>>()
-    {
-      @Override
-      public Sequence<Result<TimeseriesResultValue>> run(
-          final QueryPlus<Result<TimeseriesResultValue>> queryPlus,
-          final Map<String, Object> responseContext
-      )
-      {
-        final TimeseriesQuery query = (TimeseriesQuery) queryPlus.getQuery();
-        final Sequence<Result<TimeseriesResultValue>> baseResults = resultMergeQueryRunner.run(
-            queryPlus.withQuery(
-                queryPlus.getQuery()
-                         .withOverriddenContext(
-                             ImmutableMap.of(TimeseriesQuery.CTX_GRAND_TOTAL, false)
-                         )
-            ),
-            responseContext
-        );
+    return (queryPlus, responseContext) -> {
+      final TimeseriesQuery query = (TimeseriesQuery) queryPlus.getQuery();
+      final Sequence<Result<TimeseriesResultValue>> baseResults = resultMergeQueryRunner.run(
+          queryPlus.withQuery(
+              queryPlus.getQuery()
+                       .withOverriddenContext(
+                           ImmutableMap.of(TimeseriesQuery.CTX_GRAND_TOTAL, false)
+                       )
+          ),
+          responseContext
+      );
+
+      final Sequence<Result<TimeseriesResultValue>> finalSequence;
+
+      if (query.getGranularity().equals(Granularities.ALL) && !query.isSkipEmptyBuckets() && baseResults.toList()
```

Comment:
Calling `baseResults.toList()` here materializes the results, meaning in the case where it is _not_ empty we will actually compute it twice. And also it won't stream. To fix this, you might be able to do a trick like the one for grandTotal, where the initial sequence is lazily mapped and then concatted with a new sequence that has the extra rows you want to add.

Reply:
@gianm valid point but since the granularity is all, it is one data point anyway thus was assuming it is not too bad to call `to.List()`, what you think?

Reply:
Ah yeah, that's a good point, I forgot about short-circuiting :)

IMO, to make the code clearer & more idiot-proof it would be good to move the `toList` inside the `if` and comment that it's ok here since there's at most one granularity. Then nobody will make the same mistake in reading it that I did.

Reply:
sure will do.